making ndd.cpp 70x faster - Poly

juvi21 opened the pull request:
Hacked a little around `ndd.cpp`. I got a speedup around 70x compared to the python implementation even more to the current ndd.cpp implementation measured by a simplistic little benchmark `bench_ndd.py`

So the biggest speedups come from a little trick in Polynomial Expansion. But let’s start at the start of the func `compute_newton_coefficients`

I can store the top row of Newton coeffs directly in a single array `dd[]`. So when I label them as `dd[0..n-1]`, I can change the updated formula for the `k-th` pass to:

$$
\texttt{dd}[i]
\;=\;
\frac{\texttt{dd}[i]\;-\;\texttt{dd}[i-1]}{\,x[i]\;-\;x[i-k]\,}
\quad(\bmod\,\mathrm{MOD\_N}).
$$

That lets me iterate from n−1 down to k -> I spare some extra allocations by storing it in a compact array instead of the previous larger 2D structure.

We can now use some little tricks. Once each `dd[i]` is known, I can expand the polynomial in $O(N^2)$ time by maintaining a “factor Polynomial” $F_i(x)$

$$
P(x)
\;=\;
\sum_{i=0}^{n-1}
\texttt{dd}[i]
\;\prod_{j=0}^{i-1}(x - x[j]).
$$

So if we let $(F_i(x)$ run so that for each i we accumulate $\(\texttt{dd}[i] \times F_i(x)\)$ into the final coefficients and then update $(F_i(x)$ by multiplying by $(x - x[i])$.

So we need only one pass through the data; we can reduce Big O from $n^3$ to $n^2$ which gives quite a huge speed increase.

Also update `modInverse` to a little bit more standard EEA. The current implementation could cause some silent errors but maybe fine. Just wanted to make sure.

I experimented a little with the `calculate_divided_differences` function. But didn’t yield much improvement. I have commented it out and used the previous implementation. Just commented out `// denom = (denom + MOD_N) % MOD_N;` — shouldn’t be needed. For more details I left some comments in the function but I don’t expect a huge speed increase.

I don’t know exactly where you want to go with this repo and current state of pull request needs some cleaning. But guess the state is interesting to share.

Jackmin801 commented:
Oh nice! Yea I didnt work too much on the cpp implementation because it seemed to have a bug somewhere I couldnt quite figure out. Given a set of points, the polynomial that was obtained from the previous code sometimes didnt pass through all the points. I thought it was about an overflow somewhere so I put mods everywhere but couldnt fix it and resorted to python.

Jackmin801 commented:
Ill try out the code and if it works we can use this. Thanks for the contribution!

Jackmin801 commented:
Works great! Maybe you can clean it up abit and we can merge :)

There isnt a concrete roadmap for this repo at the moment. The near term goal is to get it packaged as a pypi library so that it can be used by generator and validator implementations. Not sure if we should include torch as a dependency. On one hand its nice to be able to directly use tensors in the library but having torch as a dependency might have its own troubles I havent fully grasped yet.

juvi21 commented:
Yeah, should have marked this as a draft :). I'll do some cleaning today or the weekend.

juvi21 commented:
Comments are now more streamlined. Old comments were comparing old to new code a lot explaining my thought process. 
I also removed the `calculate-divided_differences` function. Maybe we can keep it commented out. I moved my little simple bench into a folder called `bench/bench_ndd.py`. I also added a `.gitignore`. Probably should be in a separate pull request. Lemme know.

Jackmin801 reviewed (approved):
Nice! lgtm
